### ISSUE TITLE
QA: add SoB and WIP checks to Zuul

### ISSUE THREAD

**Issue by grahamwhaley** [kata-containers/ci#99]
In pursuit of #98 , both @ttx and I have been looking at how to configure Zuul to add SoB and WIP/label/keyword checking.
Let's open a ticket to track that progress, and share/collaborate on what we know.

## Signed-off-by / DCO checking

There is a precedent in a Zuul setup for this already. The code exists in the [ansible-network](https://github.com/ansible-network/ansible-zuul-jobs/tree/master/roles/validate-dco-license) repo, and has been [submitted](https://review.openstack.org/629565) for addition to the base Zuul libraries.

We may need to temporarily add this feature/playbook to use whilst waiting for the feature to land upstream.

## WIP/RFC label and keyword checking

Zuul has a trigger for [label/unlabel](https://zuul-ci.org/docs/zuul/admin/drivers/github.html#attr-pipeline.trigger.%3Cgithub%20source%3E.label) which I believe we can hook to do a regexp word check to block/enable PR merges.

Along with that I further suspect we can add a similar regexp check on the pull request title. Maybe we can take inspiration from the above SoB zuul example if we have to, and do it via some bash/git magic in an ansible job.

## Zuul architecture discussion

Whilst we are working on kata Zuul - I think we can start to structure the setup to present itself a little nicer to the Kata github PRs.
Right now we get the phrase 'openstack/third-party-check' for both the github app feedback and the [issue comment](https://github.com/kata-containers/proxy/pull/74#issuecomment-450180975). I think we can improve that to be much more meaningful to Kata. I propose we start breaking it down into multiple appropriately named pipes in the [zuul project config for kata](https://git.openstack.org/cgit/openstack-infra/project-config/tree/zuul.d/projects.yaml#n48). I suggest:

- kata-pre-check
  - for the static checks, like the SoB/DCO, WIP etc.
  - and I suggest we make this a template for all kata github projects similar to how the openstack projects all have a [`system-required`](https://git.openstack.org/cgit/openstack-infra/project-config/tree/zuul.d/projects.yaml#n8) template. I suggest `kata-system-required` as the template name
- kata-qa-ci
  - to cover the existing [kata QA CI test](https://github.com/kata-containers/tests/blob/master/.ci/run.sh) [playbooks](https://git.openstack.org/cgit/openstack-infra/openstack-zuul-jobs/tree/playbooks/kata-runsh)
- kata-metrics-ci
  - to cover the [kata metrics ci](https://github.com/kata-containers/tests/blob/master/.ci/run_metrics_PR_ci.sh) once we start integrating them into Zuul.

## Assignment and tracking

- [ ] Break kata zuul project into more pipelines
- [ ] Add base 'kata-system' template check
- [ ] SoB/DCO checking ([submitted](https://review.openstack.org/#/c/630294/))
- [ ] WIP/RFC label (and maybe keyword) checking

**Comment by grahamwhaley** [comment]
@ttx - this ended up slightly more verbose - I wanted to include my thoughts on structuring the kata zuul project checks slightly better whilst we are in the code.
Are there certain bits of this you have prototyped out already? Do you want to own some bits - if not, I'll forge ahead and try to update this item to keep everybody in the loop.
/cc @fungi, who might have thoughts on the rename/structuring aspects as well.

**Comment by ttx** [comment]
Please forge ahead @grahamwhaley , you are ahead of me at this point, and have the right access on the GitHub side to make fast progress. I can help grease the wheels when proposed changes hit on the Gerrit side by helping reviewing this stuff. Totally agree on customizing the way the checks present themselves, and separating the "required" checks from the more experimental jobs.

**Comment by ttx** [comment]
Ok I think I'm up to date on pipelines... sharing my thoughts for completeness. Regarding pipelines, you should define them based on what triggers them.

The first one (kata-pre-check pipeline) would be affected by PR changes ("open", "changed", "reopened") and label changes ("label") events. To replicate current behavior (WIP by comment) we could also additionally trigger it on "comment" event on specific keywords. It will run often, so we should keep long-running jobs off it.

For the second and third ones, jobs are most costly to run, so we should use a more narrow trigger. It feels like only PR changes should trigger it ("open", "changed", "reopened"). Why are you suggesting two separate pipelines (vs. a single one with two jobs) ? Do you expect the trigger to be different ? Or the reporting ?

Otherwise it should be pretty easy to add, very similar to https://git.openstack.org/cgit/openstack-infra/project-config/tree/zuul.d/pipelines.yaml#n277

Next step (for me): learn how to do the WIP ansible job

**Comment by grahamwhaley** [comment]
Sounds right to me @ttx . For the WIP check, I'm not sure we need to add a 'comment' trigger - we have not used that so far with kata, and I fear comments like 'should I make this WIP?', would trigger a block for instance ;-) And I'm not sure what comment would un-block. I'm happy to just have labels and PR subject line checks like we do at present.

My initial reason to split QA and metrics CIs into two pipelines was mostly just because historically they have been seen as separate items. I think it would be nice to see them separately in the github app section as two items (QA passed and metrics passed). afaict, pipelines appear as separate 'github app' entries (which are the first things folks look at for CI pass/fail), and 'jobs' appear as separate entries in the github comment posted by zuul (and is where the useful links to the build logs are ref'd).

There is one case where they are different - ultimately (and one of the key reasons I'm trying to move metrics CI to Zuul) we want to run the metrics CI after a PR is merged as well, and capture the results for historical data trending in the [OSF infra ELK stack ](https://docs.openstack.org/infra/system-config/logstash.html). See #60, and then there was a thread on the osf-infra email list.
Thus, we might add a 'merged' trigger to the metrics pipeline, or we might add an extra pipeline of kata-merged.

I've looked a little at the pipelines. I think we will need to check with @fungi and @clarkb that they are OK with us adding kata specific pipelines to that top level config file, and that we are doing it the correct way - just because it is pretty spartan and quite openstack specific right now. I think it is inevitable that that is going to have to happen as we get more not-openstack-specific projects coming in under the open-infra umbrella. Kata may just be one of the first, and being predominantly github based, will need to do things just a little different.

I'm just about ready to try and push the dco gerrit item. I'm expecting some feedback on that. I'll see if I can pull you in as an interested party when it is up @ttx.

**Comment by grahamwhaley** [comment]
DCO gerrit submitted. https://review.openstack.org/#/c/630294/

**Comment by fungi** [comment]
[...]
> I've looked a little at the pipelines. I think we will need to check with @fungi and @ClarkB that they are OK with us adding kata specific pipelines to that top level config file, and that we are doing it the correct way - just because it is pretty spartan and quite openstack specific right now.
[...]

With the increased complexity, I expect we're going to want to move the Kata-specific configuration into its own Zuul tenant config. We'd talked about this in the past, but the limited PoC we put together for Kata in the past used the OpenStack tenant mostly out of convenience (but also because Zuul v3 was very new at the time and we were still feeling things out so didn't want the added challenge of managing more than one tenant).

**Comment by grahamwhaley** [comment]
@fungi Sure. If/when we go that route, any thoughts on timeline etc.
I'm prone to push on with adding the SoB and WIP checks to the current kata projects as-is, as I believe we lose that functionality from pullapprove in ~17 days. Adding them into the current projects (without breaking things out into a project template or adding new kata specific pipelines) should be a fairly easy first functional step, and then we can migrate and re-structure those playbooks etc. into the new structure if/when it is set up.
Thoughts @ttx?

**Comment by ttx** [comment]
While SOB can use the existing "third-party-check" pipeline, we'll need a specific pipeline for the WIP check (one that includes a label trigger) , so I would just go and try to get that Zuul tenant set up now... We got some positive feedback from the OpenDev Zuul's team on the idea already, so it could go fast :)

It will also help with the Kata branding of the Zuul reports, to avoid the "OpenStack third-party integration" style.

**Comment by grahamwhaley** [comment]
OK, @ttx @fungi - let's try and get the zuul tenant setup if we think we can do that in time, and it necessary for the new pipeline :-)

Let me know if you need input/patches from me, and I'll hold off on my current infra zuul gerrit updates, which would be dropping the [current one](https://review.openstack.org/#/c/630294) I believe as now redundant, and would have been adding the following patch to the `project-config` repo as a first DCO check using the newly committed stdlib dco job:

```patchdiff --git a/zuul.d/projects.yaml b/zuul.d/projects.yaml
index b6edec729..48f80886a 100644
--- a/zuul.d/projects.yaml
+++ b/zuul.d/projects.yaml
@@ -49,6 +49,7 @@
     name: kata-containers/proxy
     third-party-check:
       jobs:
+        - dco-license
         - kata-runsh
         - kata-runsh-fedora-28
 
@@ -56,6 +57,7 @@
     name: kata-containers/runtime
     third-party-check:
       jobs:
+        - dco-license
         - kata-runsh
 
 - project:
```

**Comment by ttx** [comment]
@grahamwhaley to define a new tenant we'll need to add it to the main zuul config, but then point that tenant to a new git repository holding kata-specific pipelines configuration -- so I need to set that new repo up first which might take a couple of days. There is also an ongoing discussion on how to best share base jobs between tenants (which will likely result in openstack tenant zuul config being split from its base job definition), which might take a few days too.

So feel free to propose adding the dco-license job on the existing pipeline in parallel so that we can start testing that it does what we think it does :) I'll build my changes on top of yours.

**Comment by grahamwhaley** [comment]
@ttx sure. Submitted at https://review.openstack.org/#/c/630998/
I skipped templating the kata projects, as that felt like just a touch too much disruption for something we are planning to re-undo anyway.

**Comment by grahamwhaley** [comment]
hi @ttx @fungi @cboylan - any update, news, eta for the kata zuul separate tenant infra?

**Comment by ttx** [comment]
Hi @grahamwhaley, the next steps are:
1a- splitting out base jobs from openstack-specific jobs so we can set up the Kata tenant to import base jobs without importing openstack jobs. Work in progress at https://review.openstack.org/#/c/632785/
1b- set up repository for Kata tenant config -- I need to refresh https://review.openstack.org/#/c/631032/
2a- set up Kata tenant itself -- blocked on (1a)
2b- Create a WIP check job -- blocked on (1b), but someone could start working on it
3- Define pipelines and link repositories and point them to the jobs
This week we are having an offsite meeting so things are a bit slower... We should be able to make faster progress next week :)

**Comment by grahamwhaley** [comment]
Excellent, thanks for the update and links @ttx . Keep us updated, and have a good week!

**Comment by ttx** [comment]
@grahamwhaley would you have free cycles to work on a WIP-check job (2b) ? I was planning to try giving it a go next week but maybe it's better if I focus on Kata tenant  set up (1b, 2a, 3) ? Clark is helping with 1a.

**Comment by grahamwhaley** [comment]
@ttx - sure - definitely probably better that you focus on the tenant enabling. Let me have a look at what I can come up with.

**Comment by grahamwhaley** [comment]
@ttx - had a bit of a read and think around the WIP checks. Did you get anywhere btw?
- For the keywords in the PR subject line, I don't think we can see that subject line inside zuul (afaict), so I was thinking hopefully we can extract it from the merge commit in the test branch using `git` itself, a bit like the dco code does. I'll have to see how zuul does the PR merge though to see if the subject line from the PR makes it though to the repo/branch.
- For the label checks, that might actually be a touch harder. I can see how we can hook label add/remove in zuul as triggers, but we need to check the labels on every PR change, and I can't see how we can see the labels when we are in say a pull_request update - I'm not sure if they come through in the dictionary of github stuff that we can access? If they do, great, we can code something up probably. If not, then we might be a little stuffed, unless triggers in zuul can set non-transient data, in which case the label trigger could set a 'do not build' type element, and a later job could spot that and fail...

If there are any zuul github examples out there that are not of the `check` and `recheck` type triggers, let's gather them here. Something like an existing `DNM` or `WIP` check ... I'll continue searching

**Comment by grahamwhaley** [comment]
Hi @ttx I think label checks will be more useful/important than subject line checks.
Sure, I've no problem with having a pipeline just to check/process labels. I'm still not sure how we make that an 'always block' item for pull_request triggers though. I think you understand the flow I'm thinking of already, but for others and clarity, if we get the following sequence:

- WIP `label` added on github
  - label/WIP `pipeline` triggers, and would block on `WIP` label
- PR then updated on github (WIP label still in place, but rest of PR well formed)
  - label/WIP `pipeline` does not get triggered on PR update (I think), so will not block
  - pull_request/CI pipeline triggers, passes all tests, and passes the PR

Question:
- Are pipeline fails 'sticky' in zuul - that is, will a failed label/WIP pipeline block the PR until it later passes?

Thoughts on ways the WIP/label pipeline might work then:
- if it is 'sticky', that will probably do it - but I have a feeling they are not
- if the pull_request pipeline can read the labels, we will do the label check there
- if the label/WIP pipeline can set some global sticky (persistent) status, we can check that in the pull_request pipeline.
- I don't think we make the label/WIP change a pre-req step of the pull_request pipeline/job or we could only then process PR changes when a label changes, which is not our workflow (and would be horrid ;-) )
- (I thought I have another idea, but have forgotten... grrr)

My lack of Zuul knowledge is showing here ;-)

For the subject check - if I can see say a `git log --oneline master..` on the zuul branch in a slave, then we can see if the PR subject line turns up maybe in the merge commit (if there is one). We may have to do a small dance to ensure we don't get any other 'in flight PRs' that Zuul has pulled in included as well. If I can't locate such a log then I'll probably make a test item on the Zuul config/jobs and see what the logs look like..

**Comment by ttx** [comment]
For your question, it's more of a GitHub question actually. Zuul just runs the tests and reports the results. Are the previous test results sticky when a new PR is uploaded ? In Gerrit that is configurable. I don't know much of the GitHub "checks" and whether the results stick when a new PR is uploaded. If they do not stick (and GitHub resets them to "pending" until a new test result is reported) then yes I agree that we need to retrigger the pipeline and somehow access the labels (I feel like querying the GitHub API is probably the simplest, since we have the PR reference in the data passed to us). I'll do some research on that this afternoon.

**Comment by grahamwhaley** [comment]
@ttx - yeah, that occurred to me about 1h after I typed that - as the WIP would be a separate pipeline, it will appear as a separate item in github, and the status last is held by github, until it is updated (afaik). Thus, I think the WIP/label pipeline would be self contained and sticky. Let's go with that for the labels.

**Comment by ttx** [comment]
At this point I feel like it's just simpler to pull the PR using the GitHub API. We have the PR URL available to Ansible jobs as {{ zuul.change_url }} (should appear whether this is triggered by label change or PR refresh). We can feed it to a script that will apply our WIP-determining logic. For example https://gist.github.com/ttx/d961e307e04c42ea5d75b526f7f08061

**Comment by grahamwhaley** [comment]
OK. Let's spin that once the tenant/instance is up and see how it flies...

**Comment by ttx** [comment]
Alright, here is a status update...
- A repository was created to hold the Zuul configuration for the Kata tenant (zuul-config)
- The change to split the base Zuul jobs into a separate repository for use in multiple tenants is ready to merge at https://review.openstack.org/632785
- Once that merges, we can approve the change that creates the Kata tenant in the main Zuul config: https://review.openstack.org/630994
- Then define base pipelines for changes in the zuul-config repository itself: https://review.openstack.org/633938
- At that point we should be ready to add jobs for the SOB and WIP tests, and enable them through a specific pipeline triggered by GitHub PR changes and label changes. I have a PoC for the WIP test at https://gist.github.com/ttx/d961e307e04c42ea5d75b526f7f08061 -- will evolve that to support job-customizable labels and keywords.

**Comment by ttx** [comment]
All the above changes merged. Next up:

- Adding the proxy repository to the Kata tenant to test the new pipeline: https://review.openstack.org/#/c/634248/
- Create a WIP Zuul job that mimicks the PullApprove rules: https://review.openstack.org/#/c/634242/
- Defining a pipeline that triggers on label and PR changes for the WIP check, and assign the proxy repo to it: https://review.openstack.org/#/c/634245/

Those should let us test the WIP check. If that works well, we can enable it on more repositories, and move the SoB test from the openstack third-party-check pipeline to a Kata tenant pipeline.

**Comment by ttx** [comment]
The tenant is all set up now. Has SoB and WIP checks enabled for the proxy repository. You can check how it looks there. Check details link to human-friendly explanations about the check on the wiki:

* https://github.com/kata-containers/community/wiki/The-WIP-check
* https://github.com/kata-containers/community/wiki/The-SoB-check

This is ready to deploy on more repositories to replace pullapprove.

**Comment by grahamwhaley** [comment]
+1 - I've been tracking this, and it looks good to me. The [test PR over on the proxy repo](https://github.com/kata-containers/proxy/pull/141) is showing the SoB and WIP checks, and links off to the wiki pages. the wiki pages are fine (I will add a few extra hotlinks on them out to our PR requirement docs, contact channels etc.).
@ttx has raised a PR over at https://review.openstack.org/#/c/635128/ to enable the SoB and WIP checks for the other repos - we'll watch that land, and then I'll start to clear out our old pullapprove.yaml files.
we are just working out a few more details... but expect enablement etc. rsn

**Comment by ttx** [comment]
The checks are up, as mentioned in http://lists.katacontainers.io/pipermail/kata-dev/2019-February/000695.html
